### ISSUE TITLE
Markdown Support in definition floating window

### ISSUE THREAD

**Issue by phatsk** [neoclide/coc.nvim#1704]
## Result from CocInfo

```

## versions

vim version: NVIM v0.5.0-nightly
node version: v10.19.0
coc.nvim version: 0.0.76
term: iTerm.app
platform: darwin

## MessagesE325: ATTENTION

[coc.nvim] intelephense indexing ...
55 lines yanked

## Output channel: diagnostic-languageserver
Linter command: /Users/zachowen/bin/phpcs, args: ["--standard=WordPress","--report=emacs","-s","-"]
stdout: STDIN:1:1: error - Missing file doc comment (Squiz.Commenting.FileComment.Missing)
STDIN:6:12: error - Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
STDIN:27:4: error - Doc comment short description must start with a capital letter (Generic.Commenting.DocComment.ShortNotCapital)

stderr: 
match string: STDIN:1:1: error - Missing file doc comment (Squiz.Commenting.FileComment.Missing)
match result: [
  "STDIN:1:1: error - Missing file doc comment (Squiz.Commenting.FileComment.Missing)",
  "1",
  "1",
  "error",
  "Missing file doc comment (Squiz.Commenting.FileComment.Missing)"
]
match string: STDIN:6:12: error - Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
match result: [
  "STDIN:6:12: error - Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)",
  "6",
  "12",
  "error",
  "Whitespace found at end of line (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)"
]
match string: STDIN:27:4: error - Doc comment short description must start with a capital letter (Generic.Commenting.DocComment.ShortNotCapital)
match result: [
  "STDIN:27:4: error - Doc comment short description must start with a capital letter (Generic.Commenting.DocComment.ShortNotCapital)",
  "27",
  "4",
  "error",
  "Doc comment short description must start with a capital letter (Generic.Commenting.DocComment.ShortNotCapital)"
]
match string: 
match result: null


## Output channel: phpls
[Info  - 5:28:45 PM] Initialising intelephense 1.3.11
[Info  - 5:28:45 PM] Reading state from /Users/zachowen/.config/coc/extensions/coc-phpls-data/3f14a68a.
[Info  - 5:28:45 PM] Initialised in 37 ms
[Info  - 5:28:50 PM] Indexing started.
```

## Describe the bug

Documentation coming from the languageserver (Intelephense, in this case) is Markdown, however, not all of the markdown is being formatted properly in the floating window.

## Reproduce the bug

**We will close your issue when you don't provide minimal vimrc and we can't
reproduce it**

- Create file `mini.vim` with：

```vim
call plug#begin()
Plug 'neoclide/coc.nvim', {'do': 'yarn install --frozen-lockfile'}
call plug#end()

nnoremap <silent> K :call <SID>show_documentation()<CR>
```

- Start (neo)vim with command: `vim -u mini.vim`
- CocConfig settings:
```json
{
  "suggest.enablePreview": true,
  "codeLens.enable": true,
  "diagnostic.displayByAle": false,
  "languageserver": {},
  "coc.source.buffer.enable": true,
  "coc.source.around.enable": false,
  "coc.source.tag.enable": false,
  "coc.preferences.jumpCommand": "tabe",
  "coc.preferences.colorSupport": true,
  "coc.preferences.rootPatterns": [ "REDACTED", "composer.json", "wp-content", ".git" ],
  "coc.preferences.currentFunctionSymbolAutoUpdate": true,
  "diagnostic-languageserver.filetypes": {
    "php": "phpcs",
    "blade": "phpcs"
  },
  "diagnostic-languageserver.enable": true,
  "diagnostic-languageserver.linters": {
    "phpcs": {
      "command": "/Users/zachowen/bin/phpcs",
      "debounce": 100,
      "rootPatterns": [
        "centralmarket",
        "composer.json",
        "composer.lock",
        "vendor",
        ".git"
      ],
      "args": [
        "--standard=WordPress",
        "--report=emacs",
        "-s",
        "-"
      ],
      "offsetLine": 0,
      "offsetColumn": 0,
      "sourceName": "phpcs",
      "formatLines": 1,
      "formatPattern": [
        "^.*:(\\d+):(\\d+):\\s+(.*)\\s+-\\s+(.*)$",
        {
          "line": 1,
          "column": 2,
          "message": 4,
          "security": 3
        }
      ],
      "securities": {
        "error": "error",
        "warning": "warning"
      }
    }
  },
  "tsserver.implicitProjectConfig.experimentalDecorators": true,
  "solargraph.completion": true,
  "solargraph.hover": true,
  "solargraph.symbols": true,
  "solargraph.folding": true,
  "solargraph.references": true,
  "intelephense.files.associations": [
  	"*.php",
  	"*.phtml",
  	"*.blade.php"
  ],
  "intelephense.licenceKey": "REDACTED",
  "intelephense.environment.phpVersion": "7.3.0",
  "intelephense.environment.documentRoot": "REDACTED"
}
```

- Open a new PHP file with:
```php
<?php

/**
 * Foo
 *
 * @package 
 * @author Zach Owen <zach.owen@mindsize.me>
 * @version $Id$
 */
class Foo {
	/**
	 * Get something.
	 *
	 * @param int $id The ID of the thing to get.
	 *
	 * @return string
	 */
	public function get( int $id ) : string {
		return get_something( $id );
	}
}

$bar = new Foo();
$bar->get( 123 );

/**
 * get_something
 *
 * @access public
 *
 * @return string
 */
function get_something() : string {
	return 'Hey';
}
```
- Place cursor on `get` on the line with `$bar->get( 123 );`
- Press `K`

## Screenshots (optional)

Sample:
![image](https://user-images.githubusercontent.com/990374/77963175-e3cb9c00-72aa-11ea-9497-9237e379920a.png)

Expected: 
- The `__Foo::get__` would be rendered as __Foo::get__
- The `_@param_` would be rendered as _@param_

**Comment by chemzqm** [comment]
The server doesn't tell coc.nvim it's markdown format, checkout https://github.com/neoclide/coc.nvim/wiki/Debug-language-server#using-output-channel

**Comment by phatsk** [comment]
@chemzqm thanks, I'm not sure how to go about using that to setup the format for the floating window. I'll poke around and see what I can figure out - if you have any resources, that'd be great. It looks like this is what's coming back from the Language Server:

```
Result: {
    "contents": {
        "kind": "markdown",
        "value": "__Foo::get__\n\nGet something.\n\n```php\n<?php\npublic function get(int $id): string { }\n```\n\n_@param_ `int $id` — The ID of the thing to get.\n\n_@return_ `string`"
    },
    "range": {
        "start": {
            "line": 23,
            "character": 6
        },
        "end": {
            "line": 23,
            "character": 9
        }
    }
}
```

I'm not sure how to use this information to tell the floating window how to format things.

**Comment by Avi-D-coder** [comment]
@phatsk What does vs code do in this case?

**Comment by phatsk** [comment]
@Avi-D-coder honestly not sure, I don't have it installed. I can try it out and let you know.

**Comment by phatsk** [comment]
@Avi-D-coder VS Code looks like this
![image](https://user-images.githubusercontent.com/990374/78070407-b42f9900-7369-11ea-9902-6481394ba092.png)

**Comment by Avi-D-coder** [comment]
So the right thing to do would be to identify when a type signature is markdown and handle it like the rest of the window.
The trick would be minimizing false positives.

**Comment by mikehaertl** [comment]
@chemzqm I also have this issue with coc-phpls. From what I see the filetype is sent in the `kind` key of the server response as seen in the example above

```
{
    "contents": {
        "kind": "markdown",
        "value": "__Foo::get__\n\nGet something.\n\n```php\n<?php\npublic function get(int $id): string { }\n```\n\n_@param_ `int $id` — The ID of the thing to get.\n\n_@return_ `string`"
    },
```
Can't this be used by coc.nvim to set the filetype of the floating window accordingly?

If not, can you hint how the server response must look like and where this should be fixed (i.e. coc-phpls or intelephense)?

**Comment by chemzqm** [comment]
@mikehaertl The highlight should works, but we don't set filetype to avoid issues.

**Comment by mikehaertl** [comment]
@chemzqm Well, it doesn't for me. See the screenshots in the issue I've opened with intelephense here: https://github.com/bmewburn/vscode-intelephense/issues/1145 (ignore the syntax error related intendation problems).

**Comment by chemzqm** [comment]
@mikehaertl I can see the syntax highlight is works, you have highlights.

**Comment by mikehaertl** [comment]
So you say that it's expected that we see `__app\\tasks\\Task::options__` instead of __app\\\\tasks\\\\Task::options__? Is this really the best we can get? Why can't we render the floating window content as markdown to make it look pretty? Maybe I don't fully understand.

**Comment by chemzqm** [comment]
>  Is this really the best we can get?

No, PR welcome.

Because load filetype plugin could cause bad performance and bugs, coc.nvim use background neovim instance for highlights.

**Comment by mikehaertl** [comment]
> No, PR welcome.

Sorry, it was no offense. I really like coc.nvim so far. But I can't help with a PR.

I just wasn't sure if this is expected behavior. Maybe it should be added to the FAQ somehwere that markdown is displayed mostly "raw" (except for embedded code blocks) in floating popups?

**Comment by bmewburn** [comment]
@chemzqm would it help if intelephense returned `MarkedString` instead of `MarkupContent`? 

> but the server doesn't tell coc.nvim it's markdown format

`MarkedString` has a `language` property is this what is missing? If so would it still render the fenced php code blocks in addition to the markdown parts?

**Comment by chemzqm** [comment]
No difference

**Comment by bmewburn** [comment]
Thanks @chemzqm . Should coc.nvim be sending that it has markdown support then in `initialize`? Or is this controlled by coc-phpls?

```json
"hover": {
        "dynamicRegistration": true,
        "contentFormat": [
            "markdown",
            "plaintext"
        ]
    },
```

**Comment by chemzqm** [comment]
I tend to send `markdown` in `contentFormat`, the syntax might looks a bit wired but better thanm plaintext IMO, you can fork the code if you don't like it.

**Comment by mikehaertl** [comment]
@chemzqm Maybe i totally misunderstand but could you maybe give some more details why it's so difficult to render the popup content as filetype markdown? It feels like you're not in favour of doing this and I just want to understand why this is the case. If you agree that it can be improved, don't you think this is a valid feature request and worth an issue where the topic (and difficulties) can be discussed?

Maybe someone finds a solution and can come up with a PR.

**Comment by chemzqm** [comment]
> why it's so difficult to render the popup content as filetype markdown? 

I have already explained, it does syntax highlight already.

> don't you think this is a valid feature request and worth an issue where the topic (and difficulties) can be discussed?

This issue is closed, you can create new issue for feature request.

**Comment by oblitum** [comment]
the problem is that not all docs follow markdown with code fences for their docs, only a few. What comes in a floating window is in general a short declaration, or that accompanied docs in a myriad of formats. I feel like chemzqm strived for a good balance to have good enough highlight for code in that scenery, but obviously not assuming specific doc formats.

**Comment by mikehaertl** [comment]
I think there is some confusion here. The OP and I are **not** only talking about fenced code example. The *full* content from the language server is markdown:

```
"contents": {
        "kind": "markdown",
        "value": "__Foo::get__\n\nGet something.\n\n```php\n<?php\npublic function get(int $id): string { }\n```\n\n_@param_ `int $id` — The ID of the thing to get.\n\n_@return_ `string`"
    },
```


 * The `"kind": "markdown"` can be used to read the format/filetype for `value`
 * Value contains a markdown string like `__Foo::get__`
 * If you put this in a `test.md` neovim renders it fine as **Foo::get**
 * But the floating popup shows **`__Foo::get__`** instead

So I think this issue is valid and not solved yet and we want to understand the problem to find a fix.

In other words: Please forget about fenced code blocks. Yes, they have syntax highlighting. But what about the rest of the content? It's **all** markdown.

**Comment by chemzqm** [comment]
@mikehaertl There're code https://github.com/neoclide/coc.nvim/blob/master/src/model/floatBuffer.ts#L93 for highlight the unnecessry `_` and `*`,  but we can't make sure it works since people using different syntax plugins for markdown.

**Comment by chemzqm** [comment]
I don't consider it high priority since it doesn't bother me and we hope use syntax provider like tree-sister in the feature.

**Comment by oblitum** [comment]
I mentioned fenced code blocks but that's not the main point I wanted to make, that's just a detail.

The issue is that the project attempted to strive to provide reasonable code highlight, which **is not part of markdown**, but still couple that with some custom markdown highlighting. It's an ad-hoc solution to reach a good balance. If it simply picked the highlighting for markdown files, all code would not be highlighted, because programming language's syntax highlighting isn't part of markdown highlighting.

**Comment by mikehaertl** [comment]
Ok, I see the point, thanks.

But on the other hand syntax highlighting in fenced blocks works quite well  if some hints are used. At least for me with `plasticboy/vim-markdown`:

    ### Example

    ```php
    <?php
    function abc($a) { /* do something */ }
    ```

    ```javascript
    function test(a, b) { /* do something */ }
    ```

    ```html
    <h1>Some headline</h1> <span>bla bla</span>
    ```

![markdown_syntax](https://user-images.githubusercontent.com/675062/82534057-c5fd1380-9b44-11ea-82e4-d7dcd524612b.png)

I've seen that e.g. intelephense even inserts `<?php` at the beginning of the code example to make syntax highlighting work. And most libraries I use have the correct language hints.

Couldn't this become an option so that the user can decide? Something like `coc.preferences.floatFiletype`: 

* `default` gives built in highlighting like now
* `server` use whatever the language server reports in `kind` as filetype for the float window (`markdown`, `plaintext`, ...)

Probably not perfect but you get the point.

**Comment by chemzqm** [comment]
No, that doesn't make sense since there're different kinds of floating window.

**Comment by mikehaertl** [comment]
Hmm, but couldn't it be used only for the documentation floats? Or what do you mean by different kinds?

**Comment by chemzqm** [comment]
We have already support syntax highlight in fence code block of markdown, if it doesn't work for you, possible bug on your vim. The floating window could have different filetypes, we will not support override it by option.

**Comment by mikehaertl** [comment]
> We have already support syntax highlight 

Again: That's not what I and OP were asking (**Foo::get** vs **`__Foo::get__`**) but we're turning in circles now. Honestly I don't really understand the difficulties here. All would be fine if coc.nvim would just let the markdown syntax plugin let it take care of rendering. That's what the languagserver says and that's probably how vscode does it. But if you think that's not possible and coc.nvim has to use a custom markdown implementation we have to live with that.

**Comment by chemzqm** [comment]
> All would be fine if coc.nvim would just let the markdown syntax plugin let it take care of rendering

Not a option since the it would cause performance and other bugs and won't with composed filetypes, which is worse than broken highlight.

**Comment by chemzqm** [comment]
I've fixed the bold highlight in markdown, if you don't have bold text rendered, checkout your terminal preferences.

<img width="448" alt="Screen Shot 2020-05-21 at 4 58 02 PM" src="https://user-images.githubusercontent.com/251450/82542440-8fdf8400-9b84-11ea-8a9c-5aaf3c145325.png">

**Comment by mikehaertl** [comment]
@chemzqm I think the latest changes made it worse: The **backslashes** in the namespace are removed now.

* It should be `app\tasks\Task`
* It is now `apptasksTask`

![markdown_backslash](https://user-images.githubusercontent.com/675062/82817483-5e74fa00-9e9d-11ea-948c-28c835919876.png)

Before it looked like this:

![coc_underscore](https://user-images.githubusercontent.com/675062/82817623-9a0fc400-9e9d-11ea-9397-2ff844bc563b.png)

**Comment by dietrichm** [comment]
Hey

I'm also having the issue with the namespace backslashes being removed. Weirdly enough, the rest of the rendering seems fine at first sight, contrary to what @mikehaertl is showing on the first screenshot.

Anyway, I figured it must be `coc-phpls` returning double backslashes at least in the FQCN (the title). The recent fix for coc.nvim in commit 881f98ce99b5737b641690752feb11247e59da47 is removing **all** backslashes in the title. Changing L154 to the following fixes the issue for me, only replacing double backslashes by single ones:

```diff
diff --git a/src/model/floatBuffer.ts b/src/model/floatBuffer.ts
index 18464d1..c46892f 100644
--- a/src/model/floatBuffer.ts
+++ b/src/model/floatBuffer.ts
@@ -151,7 +151,7 @@ export default class FloatBuffer {
         if (doc.filetype == 'markdown') {
           // replace `\` surrounded by `__` because bug of markdown highlight in vim.
           str = str.replace(/__(.+?)__/g, (_, p1) => {
-            return `__${p1.replace(/\\/g, '')}__`
+            return `__${p1.replace(/\\\\/g, '\\')}__`
           })
         }
         lines.push(str)
```

The end result:
![Screen Shot 2020-05-29 at 10 37 44](https://user-images.githubusercontent.com/214730/83240528-cf682a80-a199-11ea-99fb-8dca5f6c2e4b.png)

I didn't want to make a PR yet but can if you want. The fix still feels "dirty" and I'm not sure if this is a correct approach that fixes all cases.

Dietrich

**Comment by mikehaertl** [comment]
> Weirdly enough, the rest of the rendering seems fine at first sight, contrary to what @mikehaertl is showing on the first screenshot.

@dietrichm Weird, indeed. Can you maybe share the relevant config options? I'm also still confused whether the float window now uses `filetype=markdown` or not, so not sure if it's relevant: But do you also use `plasticboy/vim-markdown`?

**Comment by dietrichm** [comment]
> Can you maybe share the relevant config options?

I'm pretty much on the default COC configuration settings, aside from some unrelated ones (afaik): https://github.com/dietrichm/dotfiles/blob/master/nvim/coc-settings.json

> I'm also still confused whether the float window now uses `filetype=markdown` or not

It does:

```
[Trace - 11:25:29] Received response 'textDocument/hover - (7)' in 4ms.
Result: {
    "contents": {
        "kind": "markdown",
        "value": "__League\\\\Flysystem\\\\FilesystemInterface::getMimetype__\n\nGet a file's mime-type.\n\n```php\n<?php\npublic function getMimetype($path) { }\n```\n\n_@param_ `string $path` — The path to the file.\n\n_@return_ `string|false` — The file mime-type or false on failure.\n\n_@throws_ `FileNotFoundException`"
    },
    "range": {
        "start": {
            "line": 43,
            "character": 35
        },
        "end": {
            "line": 43,
            "character": 46
        }
    }
}
```

> But do you also use `plasticboy/vim-markdown`?

I'm not.

**Comment by chemzqm** [comment]
I've made a fix, improved the replace for `\`.

**Comment by dietrichm** [comment]
> I've made a fix, improved the replace for `\`.

Thanks @chemzqm. This yields the same correct popup contents for me.

**Comment by mikehaertl** [comment]
Oh no, It still doesn't work for me. Backslashes still missing.

And it makes a difference whether I use `plasticboy/vim-markdown`.

* Without `plasticboy/vim-markdown` it looks better:
![without_markdown_plugin](https://user-images.githubusercontent.com/675062/83246308-37227380-a1a2-11ea-8a0d-da4682ac467c.png)

* With `plasticboy/vim-markdown`:
![with_markdown_plugin](https://user-images.githubusercontent.com/675062/83246322-3e498180-a1a2-11ea-83dd-428ef02a1601.png)

**Comment by dietrichm** [comment]
> Oh no, It still doesn't work for me. Backslashes still missing.

Just to be sure: are you on the master branch of coc.nvim?

**Comment by chemzqm** [comment]
@mikehaertl it's problem with detect conceal of `_`, won't fix since your syntax plugins could be using different highlight groups.

**Comment by mikehaertl** [comment]
@dietrichm Thanks for the pointer. I'm indeed using `Plug 'neoclide/coc.nvim', {'branch': 'release'}`. I want to avoid compiling js after plugin upgrade. I now wonder how I got the change that removed the backslashes some days ago: There was no new release since (at least not on github).

@chemzqm Ok, thanks for the insight. But I don't want to remove that plugin. So I will try to disable it on specific filetypes. Please let me know if you see a better fix or if you recommend a better markdown plugin.

**Comment by dietrichm** [comment]
> I now wonder how I got the change that removed the backslashes some days ago: There was no new release since (at least not on github).

The commit that introduced that change was also added or cherry-picked to the `release` branch: https://github.com/neoclide/coc.nvim/commits/release.

**Comment by chemzqm** [comment]
No, I won't do that do that for you.
I don't have cleanliness for highlight.

**Comment by mikehaertl** [comment]
@chemzqm It was only a question. But thanks anyway. If I did anything wrong - sorry for that, maybe you can enlighten me. I just try to use this plugin. 

@dietrichm Thanks for the update, that explains it then.

**Comment by ntta** [comment]
As I can see my floating window is detected as markdown, however, is there any way I can make it a markdown preview? I.e. Display the images instead of showing URLs.

![Screen Shot 2020-07-04 at 2 50 51 pm](https://user-images.githubusercontent.com/16453529/86509674-f9fa7280-be2c-11ea-8f11-2de5d9c1caa0.png)

**Comment by chemzqm** [comment]
No

**Comment by liubog2008** [comment]
![image](https://user-images.githubusercontent.com/7920667/89749435-4b67e280-dafa-11ea-86bc-bff826111564.png)

it seems that markdown highlight has some problems

@chemzqm
